adala/utils/llm.py

```diff
@@ -0,0 +1,235 @@
+import asyncio
+import instructor
+import litellm
+import traceback
+import multiprocessing as mp
+from typing import Optional, Dict, Any, List, Type, Union
+from pydantic import BaseModel, Field
+
+instructor_client = instructor.from_litellm(litellm.completion)
+async_instructor_client = instructor.from_litellm(litellm.acompletion)
+
+
+class LLMResponse(BaseModel):
+    data: Dict = Field(default_factory=dict)
+    adala_error: bool = Field(default=False, serialization_alias='_adala_error')
+    adala_message: str = Field(default=None, serialization_alias='_adala_message')
+    adala_details: str = Field(default=None, serialization_alias='_adala_details')
```

Comment:
definitely appreciate using a typed object here instead of a dataframe :+1:

Reply:
can fields within `data` also be added to the schema here?

Reply:
different objects for constrained and unconstrained generation might be clearer

Reply:
> can fields within data also be added to the schema here?

no, because they are user-defined, we don't know them in advance therefore can not specify in model fields

Reply:
> different objects for constrained and unconstrained generation might be clearer

good idea, I can support that, however, I don't see any reason why we need unconstrained generation at all, and it can be removed in the future versions

Reply:
resolved with https://github.com/HumanSignal/Adala/pull/162#discussion_r1692436952